Migrate to credentials.toml for auth token

wfchandler opened the pull request:
With Release 9 we removed support for `hosts.toml` files in `oxide.rs`.

Add support for the new `credentials.toml` format and warn users that `hosts.toml` is deprecated.

wfchandler committed:
Migrate to credentials.toml for credentials

With Release 9 we removed support for `hosts.toml` files in `oxide.rs`.

Add support for the new `credentials.toml` format and warn users that
`hosts.toml` is deprecated.

wfchandler commented:
Maybe it's better to just pull the bandaid off and make a hard cutover to the new format, but I went with the soft touch as a starting place.

```
$ omicron-stress --host-uri https://oxide.sys.r3.oxide-preview.com --hosts-toml-dir ~/.config/oxide
2024-08-12T18:34:19.148450Z  INFO omicron_stress::client: 117: Nexus URI host=https://oxide.sys.r3.oxide-preview.com
2024-08-12T18:34:19.148654Z  INFO omicron_stress::client: 145: reading credentials from /Users/wfc/.config/oxide
2024-08-12T18:34:19.150353Z  INFO omicron_stress::client: 147: attempting to read token from credentials.toml
2024-08-12T18:34:19.155679Z  INFO omicron_stress: 35: Checking for existing stress project
2024-08-12T18:34:19.626419Z  INFO omicron_stress: 37: Project already exists
```

hawkw commented on README.md:
```diff
@@ -21,7 +21,8 @@ in order):
 (again evaluated in order):
 
-- A `hosts.toml` file stored in one of the following locations:
+- A `credentials.toml` or the deprecated `hosts.toml` file stored in one of the following locations:
```
since the  `hosts.toml` file _is_ still supported, perhaps it should still be mentioned here, but explicitly labeled as deprecated? not a big deal

ahl replied:
@hawkw I'd love to burn our boats on this one and push everyone to the new method. What do you think?

wfchandler replied:
I think I must be misunderstanding, the comment does mention that `hosts.toml` is deprecated. Are you thinking a separate line that makes it easier to notice that we're moving away from it?

hawkw replied:
Never mind, I can't read. Carry on!

hawkw commented on src/client.rs:
```diff
@@ -22,11 +22,87 @@ struct Host {
+        match (creds.exists(), hosts.exists()) {
+            (true, _) => "credentials.toml",
+            (_, true) => "hosts.toml",
+            (false, false) => "no config found",
```
This kind of makes me feel like the function should probably return an `Option` (or  `Result`?) rather than a string for "no config found"?

wfchandler replied:
I think this was a sign that we needed tighter logic around the type of config found. I've added an enum to capture this.

hawkw reviewed (commented):
thanks for doing this, it would've made my life a lot easier a week ago! :)

wfchandler committed:
Validate config type immediately

Rather than lazily checking for the config type found when reading it,
check immediately when constructing `LoginConfig`. Add a `ConfigType`
enum to capture which type of config will be used.

wfchandler commented:
Thanks for the review @hawkw! Ready for another look.

jmpesp commented on src/client.rs:
```diff
@@ -52,33 +146,31 @@ pub fn get_client(config: &crate::config::Config) -> Result<oxide::Client> {
-    // Attempt to read `hosts.toml` and extract a token from it. If this fails
-    // for any reason (`hosts.toml` not found or malformed, or no search path
+    // Attempt to read credentials config and extract a token from it. If this fails
+    // for any reason (`credentials/hosts.toml` not found or malformed, or no search path
```
`credentials/hosts.toml` reads like a path to me, but I think the intention is `credentials.toml or hosts.toml`?

wfchandler replied:
That's right, fixed.

jmpesp commented on src/config.rs:
```diff
@@ -47,6 +47,9 @@ pub struct Config {
     pub hosts_toml_dir: Option<PathBuf>,
 
+    #[arg(long)]
+    pub credentials_toml_dir: Option<PathBuf>,
```
This should also be commented, similar to what `hosts_toml_dir` has

wfchandler replied:
Whoops, thanks!

jmpesp commented on src/client.rs:
```diff
@@ -22,11 +22,98 @@ struct Host {
-    dir.push("hosts.toml");
-    let hosts = std::fs::read_to_string(dir)?;
-    Ok(toml::from_str(&hosts)?)
+/// The contents of an Oxide CLI `credentials.toml` file.
```
Not blocking: a small example of what the serialized file looks like would be nice, or example field values.

wfchandler replied:
Makes sense, added.

wfchandler committed:
Fixup doc comments and add help text
